### PR TITLE
*: Add progress notify request watch request

### PR DESCRIPTION
**Pull request by jpbetz**
Add `RequestProgress` function to `Watcher` interface to request a progress notify response in all watch channels.

That can be used to check if a watcher is "caught up" to a particular revision, which is useful when determining if a watch cache is stale by comparing the progress revision to the revision returned in a quorum read.

Note that this also changes how progress notify events are communicated. They are now broadcast to all watch channels.

Fixes https://github.com/coreos/etcd/issues/9855

Example etcdctl usage:

```
ETCDCTL_API=3 bin/etcdctl watch --interactive
watch a
progress
progress notify: 14
watch b
progress
progress notify: 14
progress notify: 14 # since progress notify is broadcast to both watch channels
# In a separate terminal: ETCDCTL_API=3 bin/etcdctl put c 2
progress
progress notify: 15
...
```

### CODE REVIEW

**Comment by jpbetz** [comment]
cc @wenjiaswe

File: etcdctl/ctlv3/command/watch_command.go [hunk 1]
```diff
@@ -101,27 +101,33 @@ func watchInteractiveFunc(cmd *cobra.Command, osArgs []string, envKey, envRange
...
+			}
 
-		if args[0] != "watch" {
+			ch, err := getWatchChan(c, watchArgs)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Invalid command %s (%v)\n", l, err)
+				continue
+			}
+			go printWatchCh(c, ch, execArgs)
+		case "progress":
```
Comment (gyuho) [inline etcdctl/ctlv3/command/watch_command.go]: Q. Does this work? Think watch streams are not shared in etcdctl? So, this progress request would be no-op?

File: etcdserver/etcdserverpb/rpc.proto [hunk 2]
```diff
@@ -367,6 +367,9 @@ message ResponseHeader {
   // member_id is the ID of the member which sent the response.
   uint64 member_id = 2;
   // revision is the key-value store revision when the request was applied.
+  // For watch progress responses, the header.revision indicates progress. All future events
+  // recieved in this stream are guarenteed to have a higher revision number than the
```
Comment (gyuho) [inline etcdserver/etcdserverpb/rpc.proto]: `s/guarenteed/guaranteed/` :)

File: etcdctl/ctlv3/command/watch_command.go (hunk 1 above)
Comment (jpbetz) [reply etcdctl/ctlv3/command/watch_command.go]: Hm. I'm not familiar with that aspect of etcdctl.  Here's a sample run I just did:
```
ETCDCTL_API=3 bin/etcdctl watch --interactive
watch a
progress
progress notify: 14
watch b
progress
progress notify: 14
```

Is there a case you can think of that might exercise a problem with the separate watch streams?

Comment (gyuho) [reply etcdctl/ctlv3/command/watch_command.go]: @jpbetz Oh, actually you are right. This is for interactive watch command, so requests are sharing the stream. I was thinking of just regular watch command, and a separate progress command.

File: etcdserver/etcdserverpb/rpc.proto (hunk 2 above)
Comment (jpbetz) [reply etcdserver/etcdserverpb/rpc.proto]: Fixed and re-ran genproto.

File: etcdctl/ctlv3/command/watch_command.go (hunk 1 above)
Comment (jpbetz) [reply etcdctl/ctlv3/command/watch_command.go]: @gyuho I can work on adding more edge cases to the `TestWatchRequestProgress` clientv3/integration test that I added in this PR.  I currently just does:  (1) start a watch (2) put a k/v (3) verify the k/v is recieved over the watch (4) request progress (5) check that progress is received over the watch.

Corner cases might be:
- No watches
- More than one watch
- Put a key not visible to the watch, ensure progress is incremented

If you think of any others, let me know.

File: clientv3/watch.go [hunk 3]
```diff
@@ -608,7 +662,28 @@ func (w *watchGrpcStream) dispatchEvent(pbresp *pb.WatchResponse) bool {
 		Canceled:        pbresp.Canceled,
 		cancelReason:    pbresp.CancelReason,
 	}
-	ws, ok := w.substreams[pbresp.WatchId]
+
+	if wr.IsProgressNotify() {
```
Comment (jpbetz) [inline clientv3/watch.go]: @gyuho Since progress notify is global, does it make sense to broadcast the watch response to all channel substreams? This will change the behavior of existing progress notifies as well.

File: etcdctl/ctlv3/command/watch_command.go (hunk 1 above)
Comment (jpbetz) [reply etcdctl/ctlv3/command/watch_command.go]: I modified the watch stream publisher to send the progress notify responses on all watch channels, the result is etcdctl now does:

```
ETCDCTL_API=3 bin/etcdctl watch --interactive
watch a
progress
progress notify: 14
watch b
progress
progress notify: 14
progress notify: 14
```

File: clientv3/watch.go (hunk 3 above)
Comment (gyuho) [reply clientv3/watch.go]: Yes, since Progress Notify request is sent with no specific watch ID, it makes sense to dispatch globally.

**Comment by jpbetz** [comment]
@gyuho @xiang90 This is ready for review.

File: Documentation/dev-guide/interacting_v3.md [hunk 4]
```diff
@@ -355,6 +355,23 @@ foo         # key
 bar_latest  # value of foo key after modification
 ```
 
+## Watch progress
+
+Applications may want to check the progress of a watch to determine how up-to-date the watch stream is. For example, if a watch is used to update a cache, it can be useful to know if the cache is stale compared to the revision from a qoroum read.
```
Comment (gyuho) [inline Documentation/dev-guide/interacting_v3.md]: `s/qoroum/quorum/` :)

File: clientv3/integration/watch_test.go [hunk 5]
```diff
@@ -582,6 +582,78 @@ func testWatchWithProgressNotify(t *testing.T, watchOnPut bool) {
...
+
+			// verify all watch channels receive a progress notify
+			for _, rch := range watchChans {
+				select {
+				case resp := <-rch:
+					if !resp.IsProgressNotify() {
+						t.Fatalf("expected resp.IsProgressNotify() == true")
+					}
+					if resp.Header.Revision != 3 {
+						t.Fatalf("resp.Header.Revision expected 2, got %d", resp.Header.Revision)
```
Comment (gyuho) [inline clientv3/integration/watch_test.go]: `s/resp.Header.Revision expected 2/resp.Header.Revision expected 3/`?

File: Documentation/dev-guide/interacting_v3.md (hunk 4 above)
Comment (gyuho) [reply Documentation/dev-guide/interacting_v3.md]: Can we also clarify this revision in watch response to progress notify request *is* revision from the local node that the watch stream is connected to? Since quorum read cannot be returned from the partitioned node, mention something like "stale compared to the revision fetched from a healthy cluster (e.g. header in quorum read)".

File: etcdctl/ctlv3/command/watch_command.go (hunk 1 above)
Comment (gyuho) [reply etcdctl/ctlv3/command/watch_command.go]: @jpbetz I tried those corner cases with this branch, and confirm that it's covered. I can't think of any, for now.
Thanks!

File: clientv3/watch.go [hunk 6]
```diff
@@ -72,6 +72,9 @@ type Watcher interface {
 
 	// Close closes the watcher and cancels all watch requests.
 	Close() error
+
+	// RequestProgress requests a progress notify response be sent in all watch channels.
+	RequestProgress(ctx context.Context) error
```
Comment (xiang90) [inline clientv3/watch.go]: Move this above Close method? Also rename this to just Progress?

Comment (xiang90) [reply clientv3/watch.go]: Actually, after second thought RequestProgress is better.

Comment (jpbetz) [reply clientv3/watch.go]: Thanks @xiang90, I've moved above `Close()`.

File: clientv3/integration/watch_test.go (hunk 5 above)
Comment (jpbetz) [reply clientv3/integration/watch_test.go]: Done. Thanks for the catch.

File: Documentation/dev-guide/interacting_v3.md (hunk 4 above)
Comment (jpbetz) [reply Documentation/dev-guide/interacting_v3.md]: Ah, yes, I should be more explicit about that revision is returned. Fixing.

File: clientv3/watch.go (hunk 3 above)
Comment (xiang90) [reply clientv3/watch.go]: @jpbetz @gyuho 

this changes the previous behavior of watcher. the notification is per watch creation not per stream.

**Comment by xiang90** [comment]
@jpbetz We still need to preserve the old behavior of progress notification.

**Comment by gyuho** [comment]
@xiang90 Good point.

Problem with current Protocol Buffer definition is we do not differentiate regular progress notify *and* the one triggered via request progress. Also another problem is we *do not* expose watch response ID, that can be used to pinpoint which stream to receive progress notify. So, it makes sense to broadcast the response for manually triggered progress request. But, it breaks the behavior of periodic progress notify response. Maybe add an extra field to denote that this response is requested by the request progress request?

**Comment by gyuho** [comment]
@jpbetz 

> This might be as simple as something like if `pbresp.WatchId == 0 && wr.IsProgressNotify() {`

Sounds good. Adding extra check `pbresp.WatchId == 0` before broadcasting would be simpler.

**Comment by jpbetz** [comment]
To preserve the previous behavior I've assigned -1 watch IDs to watch responses for `RequestProgress`.  0 did not work correctly since watch IDs are zero indexed (the 1st watch channel is assigned `WatchId` 0).

This makes it clear that the response is not associated with any particular watch id, which seems helpful in preventing from the watch response from being mis-interpreted.  Since it's not obvious that it should be broadcast to all channels, I've added a comment explaining that.

**Comment by gyuho** [comment]
Can we fix the CI?

> clientv3/watch.go.671: // watchIds are zero indexed, so RequestNotify watch responses are assigned watchId -1 to (spell: watchIds -> watch Ids?)

We just need to whitelist `watchIds` in `.words` file :)
